R/learner_BART_surv_bart.R

```diff
@@ -0,0 +1,293 @@
+          offset     = p_dbl(default = NULL, special_vals = list(NULL), tags = "train"),
+          ntree      = p_int(default = 50L, lower = 1L, tags = "train"),
+          numcut     = p_int(default = 100L, lower = 1L, tags = "train"),
+          ndpost     = p_int(default = 1000L, lower = 1L, tags = "train"),
+          nskip      = p_int(default = 250L, lower = 0L, tags = "train"),
+          keepevery  = p_int(default = 10L, lower = 1L, tags = "train"),
+          printevery = p_int(default = 100L, lower = 1L, tags = "train"),
+          seed       = p_int(default = 99L, tags = "train"),
+          mc.cores   = p_int(default = 2L, lower = 1L, tags = c("train", "predict")),
+          nice       = p_int(default = 19L, lower = 0L, upper = 19L, tags = c("train", "predict")),
+          openmp     = p_lgl(default = TRUE, tags = "predict"),
+          quiet      = p_lgl(default = TRUE, tags = "predict")
+        )
+
+        # custom defaults
+        param_set$values = list(mc.cores = 1, quiet = TRUE)
+
+        super$initialize(
+          id = "surv.bart",
+          packages = "BART",
+          feature_types = c("logical", "integer", "numeric", "factor"),
+          predict_types = c("crank", "distr"),
+          param_set = param_set,
+          properties = c("importance", "missings"),
+          man = "mlr3extralearners::mlr_learners_surv.bart",
+          label = "Bayesian Additive Regression Trees"
+        )
+      },
+
+      #' @description
+      #' Two types of importance scores are supported:
+      #' 1. The mean selection probability of each feature in the trees,
+      #' extracted from the slot `varprob.mean`.
+      #' If `sparse = FALSE` (default), this is a fixed constant.
+      #' Recommended to use this option when `sparse = TRUE`.
+      #' 2. The observed count of each feature in the trees, extracted from the
+      #' slot `varcount.mean`.
+      #' This is the default importance scores.
+      #'
+      #' In both cases, higher values signify more important variables.
+      #'
+      #' @param type Can be either `count` or `prob`.
+      #' @return Named `numeric()`.
+      importance = function(type = 'count') {
+        if (!type %in% c('prob', 'count')) stopf("type can be only 'prob' or 'count'")
+
+        if (is.null(self$model)) {
+          stopf("No model stored")
+        }
+
+        if (type == 'prob') {
+          sort(self$model$varprob.mean[-1], decreasing = T)
+        } else {
+          sort(self$model$varcount.mean[-1], decreasing = T)
+        }
+      }
+    ),
+
+    private = list(
+      .train = function(task) {
+        pars = self$param_set$get_values(tags = "train")
+
+        x_train = as.data.frame(task$data(cols = task$feature_names))
+        times   = task$truth()[,1]
+        delta   = task$truth()[,2] # delta => status
+
+        # need these for predict
+        self$state$train_data = list(x_train = x_train, times = times, delta = delta)
```

Comment:
the `.train()` function should return a list that contains the model and this data as well. otoh storing the x_train costs a lot of memory for large datasets so it should be possible to switch it off